File: 卡码/代码随想录/1/11-图论/C++/graph/dfs/LeetCode_0841.cpp
//841. 钥匙和房间
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
	vector<bool> visited;
public:
	void dfs(vector<vector<int>>& rooms,int key){
		if(visited[key]) return;
		visited[key]=true;
		vector<int> keys=rooms[key];
		for(int key:keys){
			dfs(rooms,key);
		}
	}
	bool canVisitAllRooms(vector<vector<int>>& rooms) {
		visited.resize(rooms.size(),false);
		dfs(rooms,0);
		return count(visited.begin(),visited.end(),false)==0;
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/1/11-图论/Java/src/com/kurumi/dfs/LeetCode_0841.java
//841. Ô¿³×ºÍ·¿¼ä
package com.kurumi.dfs;

import java.util.List;

public class LeetCode_0841 {

}
class Solution {
	boolean[] visited;
	void dfs(List<List<Integer>> rooms,int key) {
		if(visited[key]) return;
		visited[key]=true;
		List<Integer> keys = rooms.get(key);
		for(int k:keys) {
			if(!visited[k]) dfs(rooms,k);
		}
	}
  public boolean canVisitAllRooms(List<List<Integer>> rooms) {
  	visited=new boolean[rooms.size()];
  	dfs(rooms,0);
  	for(boolean visit:visited) {
  		if(!visit) return false;
  	}
  	return true;
  }
}